sputnikdao2/src/proposals.rs

```diff
@@ -0,0 +1,430 @@
+        receiver_id: AccountId,
+        amount: U128,
+    },
+    /// Mints new tokens inside this DAO.
+    Mint { amount: U128 },
+    /// Burns tokens inside this DAO.
+    Burn { amount: U128 },
+    /// Add new bounty.
+    AddBounty { bounty: Bounty },
+    /// Indicates that given bounty is done by given user.
+    BountyDone {
+        bounty_id: u64,
+        receiver_id: AccountId,
+    },
+    /// Just a signaling vote, with no execution.
+    Vote,
+}
+
+impl ProposalKind {
+    /// Returns label of policy for given type of proposal.
+    pub fn to_policy_label(&self) -> &str {
+        match self {
+            ProposalKind::ChangeConfig { .. } => "config",
+            ProposalKind::ChangePolicy { .. } => "policy",
+            ProposalKind::AddMemberToRole { .. } => "add_member_to_role",
+            ProposalKind::RemoveMemberFromRole { .. } => "remove_member_from_role",
+            ProposalKind::FunctionCall { .. } => "call",
+            ProposalKind::UpgradeSelf { .. } => "upgrade_self",
+            ProposalKind::UpgradeRemote { .. } => "upgrade_remote",
+            ProposalKind::Transfer { .. } => "transfer",
+            ProposalKind::Mint { .. } => "mint",
+            ProposalKind::Burn { .. } => "burn",
+            ProposalKind::AddBounty { .. } => "add_bounty",
+            ProposalKind::BountyDone { .. } => "bounty_done",
+            ProposalKind::Vote => "vote",
+        }
+    }
+}
+
+/// Votes recorded in the proposal.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug)]
+#[serde(crate = "near_sdk::serde")]
+pub enum Vote {
+    Approve = 0x0,
+    Reject = 0x1,
+    Remove = 0x2,
+}
+
+impl From<Action> for Vote {
+    fn from(action: Action) -> Self {
+        match action {
+            Action::VoteApprove => Vote::Approve,
+            Action::VoteReject => Vote::Reject,
+            Action::VoteRemove => Vote::Remove,
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Proposal that are sent to this DAO.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[cfg_attr(feature = "test", derive(Debug))]
+#[serde(crate = "near_sdk::serde")]
+pub struct Proposal {
+    /// Original proposer.
+    pub proposer: AccountId,
+    /// Description of this proposal.
+    pub description: String,
+    /// Kind of proposal with relevant information.
+    pub kind: ProposalKind,
+    /// Current status of the proposal.
+    pub status: ProposalStatus,
+    /// Count of votes per decision: yes / no / spam.
+    pub vote_counts: [Balance; 3],
+    /// Map of who voted and how.
+    pub votes: HashMap<AccountId, Vote>,
+    /// Submission time (for voting period).
+    pub submission_time: WrappedTimestamp,
+}
+
+impl Proposal {
+    /// Update count and map of votes.
+    pub fn update_votes(&mut self, account_id: AccountId, vote: Vote, amount: Balance) {
+        self.vote_counts[vote.clone() as usize] += amount;
+        self.votes.insert(account_id, vote);
```

Comment:
Probably want to check that you haven't voted.
```suggestion
        assert!(self.votes.insert(account_id, vote).is_none(), "ERR_ALREADY_VOTED");
```